### PR TITLE
Implement main mario animations

### LINKED ISSUE #15: Implement all of Mario's movement animations and death animation
### [Footage](https://imgur.com/a/KPZqeDD)

Small & Super
- Idle - OK
- Walk - OK
- Run - OK
- Skid - OK
- Jump - OK (no charge)
- Run jump - OK
- Spin jump - OK
- Look up - OK
- Duck - OK
- Fall - OK
- Climbing vines - OK

Death - OK
Grow up - OK
Grow down - OK

Riding Yoshi and holding animations will be done later.

### PR DESCRIPTION
**Pull request by sandenson**
### [Footage](https://imgur.com/a/KPZqeDD)

Small & Super
- Idle - OK
- Walk - OK
- Run - OK
- Skid - OK
- Jump - OK (no charge)
- Run jump - OK
- Spin jump - OK
- Look up - OK
- Duck - OK
- Fall - OK
- Climbing vines - OK

Death - OK
Grow up - OK
Grow down - OK

Riding Yoshi and holding animations will be done later.

### CODE REVIEW

**Comment by sandenson** [comment]
Small & Super
- Idle - OK
idle
- Walk - OK (pretty laggy right now, needs fixing [later](https://github.com/sandenson/samba-mario-world/issues/16))
walk
- Run - OK (pretty laggy right now, needs fixing [later](https://github.com/sandenson/samba-mario-world/issues/16))
run
- Skid - OK
skid
- Jump - OK (no charge)
jump
- Run jump - OK (stopped working; will look into that [later](https://github.com/sandenson/samba-mario-world/issues/16))
- Spin jump - OK
spin_jump
- Look up - OK
look_up
- Duck - OK
duck
- Fall - OK
fall
- Climbing vines - OK

Death - OK
death
Grow up - OK
grow_up
Grow down - OK
grow_down

Riding Yoshi and holding animations will be done later.

File: assets/scripts/camera/Camera.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  public update(dt: number): void {
+    if (this.player.isValid) {
+      const targetPosition = this.player.x;
+
+      const boundaryX = this.node.parent.convertToWorldSpaceAR(this.boundary.getPosition()).x - this.boundary.width / 2;
+      const finalBoundaryX =
+        this.boundary.width + boundaryX - this.node.parent.convertToNodeSpaceAR(this.node.getPosition()).x / 6;
+      // const finalBoundaryX = this.boundary.width * this.boundary.scaleX - this.node.parent.width;
+      // eslint-disable-next-line no-console
+      // console.log(`${boundaryX} ${this.node.x}`);
```
Comment (Manuel-Antunes) [inline assets/scripts/camera/Camera.ts]:
```suggestion
```
remove useless commits

File: assets/scripts/characters/Character.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+  public set isJumping(value: boolean) {
+    this._isJumping = value;
+  }
+
+  public facing: DIRECTION = DIRECTIONS.right;
+
+  public abstract jumpForce = 0;
+
+  public abstract moveForce = 0;
```
Comment (Manuel-Antunes) [inline assets/scripts/characters/Character.ts]: why abstract ?

File: assets/scripts/characters/DIRECTION.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+  up: { x: 0, y: 1 },
+  down: { x: 0, y: -1 },
+} as const;
+
+interface DIRECTION {
+  x: number;
+  y: number;
+}
+
+export default DIRECTION;
```
Comment (Manuel-Antunes) [inline assets/scripts/characters/DIRECTION.ts:13]: verry nice!

File: assets/scripts/characters/IDyingCharacter.ts [hunk 4]
```diff
@@ -0,0 +1,3 @@
+export default interface IDyingCharacter {
```
Comment (Manuel-Antunes) [inline assets/scripts/characters/IDyingCharacter.ts:1]: interface segration principle, nice!

File: assets/scripts/characters/mario/Mario.ts [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+  }
+
+  // for now, this setter is pretty much useless, but it'll be needed later
+  public set isImmortal(value: boolean) {
+    if (value !== this.isImmortal) {
+      this._isImmortal = value;
+    }
+  }
+
+  private _isSuper = false;
```
Comment (Manuel-Antunes) [inline assets/scripts/characters/mario/Mario.ts:55]: I know at this level you can only get bigger, but, i think i don't like this way to work in the code using a single variable.

File: assets/scripts/characters/mario/Mario.ts [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+export default class Mario extends Character implements IDyingCharacter {
+  private isAlive = true;
+
+  public jumpForce = 9000;
+
+  public moveForce = 600;
+
+  public maxSpeed = 200;
+
+  public walkingMaxSpeed = 200;
```
Comment (Manuel-Antunes) [inline assets/scripts/characters/mario/Mario.ts]: Why not property?

File: assets/scripts/characters/mario/MarioStateMachine.ts [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+          .call(() => this.actor.node.destroy())
+          .start();
+      }
+    }
+  }
+
+  public onAnimationEnd(name: string): void {
+    if (name === 'grow_up') {
+      this.actor.canMove = true;
+    }
```
Comment (Manuel-Antunes) [inline assets/scripts/characters/mario/MarioStateMachine.ts:39]: The player don't stop to walk, all the game pauses when mario get a power

File: assets/scripts/controllers/KeyboardController.ts [hunk 8]
```diff
@@ -0,0 +1,20 @@
+// Learn TypeScript:
+//  - https://docs.cocos.com/creator/manual/en/scripting/typescript.html
+// Learn Attribute:
+//  - https://docs.cocos.com/creator/manual/en/scripting/reference/attributes.html
+// Learn life-cycle callbacks:
+//  - https://docs.cocos.com/creator/manual/en/scripting/life-cycle-callbacks.html
```
Comment (Manuel-Antunes) [inline assets/scripts/controllers/KeyboardController.ts]:
```suggestion
```
remove useless comments in all files

**Review by Manuel-Antunes — CHANGES_REQUESTED** [review]
please, fix all issues to proceed

File: assets/scripts/characters/DIRECTION.ts (hunk 3 above)
Comment (sandenson) [reply assets/scripts/characters/DIRECTION.ts:13]: Bela pica

File: assets/scripts/characters/IDyingCharacter.ts (hunk 4 above)
Comment (sandenson) [reply assets/scripts/characters/IDyingCharacter.ts:1]: Bela pica

Comment (sandenson) [reply assets/scripts/characters/IDyingCharacter.ts:1]: Thanks.

File: assets/scripts/characters/mario/Mario.ts (hunk 5 above)
Comment (sandenson) [reply assets/scripts/characters/mario/Mario.ts:55]: In that case, do you have any recommendations?

File: assets/scripts/characters/mario/Mario.ts (hunk 6 above)
Comment (sandenson) [reply assets/scripts/characters/mario/Mario.ts]: I don't want this to be a property. Mario won't change any more than what's implemented in the code.

File: assets/scripts/characters/mario/MarioStateMachine.ts (hunk 7 above)
Comment (sandenson) [reply assets/scripts/characters/mario/MarioStateMachine.ts:39]: This is a temporary measure, as there isn't a game controller yet. I plan to implement one in the next PR I'll work on.

File: assets/scripts/controllers/KeyboardController.ts (hunk 8 above)
Comment (sandenson) [reply assets/scripts/controllers/KeyboardController.ts]: Done.

File: assets/scripts/characters/Character.ts (hunk 2 above)
Comment (sandenson) [reply assets/scripts/characters/Character.ts]: I'd forgotten about that already. It's been removed now.

**Review by Manuel-Antunes — APPROVED** [review]
Handsome Penis

**Comment by sandenson** [comment]
> Handsome Penis

😳

**Pull request merged**